miniapps/esl-translate.man
.TH "esl-translate" 1  "@RELEASEDATE@" "@PACKAGE@ @RELEASE@" "@PACKAGE@ Manual"

.SH NAME
.TP
esl-translate - translate DNA sequence in six frames into individual ORFs

.SH SYNOPSIS
.B esl-translate
.I [options]
.I seqfile


.SH DESCRIPTION

.pp
Given a 
.I seqfile 
containing DNA or RNA sequences,
.B esl-translate
outputs a six-frame translation of them as individual open reading
frames in FASTA format.

.pp
By default, only open reading frames greater than 20aa are reported. 
This minimum ORF length can be changed with the
.I -l 
option.

.pp
By default, no specific initiation codon is required, and any amino acid can start an open reading frame.
This is because
.I esl-translate
may be used on sequence fragments, eukaryotic genes with introns, or other
cases where we
do not want to assume that ORFs are complete coding regions.
This behavior can be changed. With the 
.I -m 
option, ORFs start with an initiator AUG Met. With the
.I -M
option, ORFs start with any of the initiation codons allowed by the
genetic code. For example, the "standard" code (NCBI transl_table 1) 
allows AUG, CUG, and UUG as initiators. When
.I -m
or
.I -M
are used, an initiator is always translated to Met (even if the initiator
is something like UUG or CUG that doesn't encode Met as an elongator).

.pp
If
.I seqfile
is - (a single dash), input is read from the stdin pipe. This
(combined
with the output being a standard FASTA file) allows
.B esl-translate 
to be used in command line incantations.
If
.I seqfile
ends in .gz, it is assumed to be a gzip-compressed file, and 
Easel will try to read it as a stream from
.B gunzip -c.



.SH OUTPUT FORMAT

.pp


The output FASTA name/description line contains information about the
source and coordinates of each ORF. Each ORF is named 
.B orf1,
etc., with numbering starting from 1, in order of their start position
on the top strand followed by the bottom strand.  The rest of the
FASTA name/desc line contains 4 additional fields, followed by the
description of the source sequence:

.TP
.B source=<s>
<s> is the name of the source DNA/RNA sequence.

.TP
.B coords=<start>..<end>
Coords, 1..L, for the translated ORF in a source DNA sequence of
length L. If start is greater than end, the ORF is on the bottom
(reverse complement) strand. The start is the first nucleotide of the
first codon; the end is the last nucleotide of the last codon. The
stop codon is not included in the coordinates (unlike in CDS
annotation in GenBank, for example.)

.TP
.B length=<n>
Length of the ORF in amino acids.

.TP
.B frame=<n>
Which frame the ORF is in. Frames 1..3 are the top strand; 4..6 are the
bottom strand. Frame 1 starts at nucleotide 1. Frame 4 starts at
nucleotide L.



.SH ALTERNATIVE GENETIC CODES

.pp
By default, the "standard" genetic code is used (NCBI transl_table 1). 
Any NCBI genetic code transl_table can be selected with the
.I -c 
option, as follows:

.TP 
.B 1 
Standard
.TP
.B 2 
Vertebrate mitochondrial
.TP
.B 3
Yeast mitochondrial
.TP
.B 4 
Mold, protozoan, coelenterate mitochondrial; Mycoplasma/Spiroplasma
.TP
.B 5 
Invertebrate mitochondrial
.TP
.B 6 
Ciliate, dasycladacean, Hexamita nuclear
.TP
.B  9 
Echinoderm and flatworm mitochondrial
.TP
.B 10 
Euplotid nuclear
.TP
.B 11
Bacterial, archaeal; and plant plastid
.TP
.B 12 
Alternative yeast
.TP
.B 13 
Ascidian mitochondrial
.TP
.B 14 
Alternative flatworm mitochondrial
.TP
.B 16 
Chlorophycean mitochondrial
.TP
.B 21 
Trematode mitochondrial
.TP
.B 22 
Scenedesmus obliquus mitochondrial
.TP
.B 23 
Thraustochytrium mitochondrial
.TP
.B 24 
Pterobranchia mitochondrial
.TP
.B 25 
Candidate Division SR1 and Gracilibacteria


.PP
As of this writing, more information about the genetic codes in the
NCBI translation tables is at 
.I http://www.ncbi.nlm.nih.gov/Taxonomy/ 
at a link titled
.I Genetic codes.

.SH IUPAC DEGENERACY CODES IN DNA 

.pp
DNA sequences may contain IUPAC degeneracy codes, such as N, R, Y,
etc. If all codons consistent with a degenerate codon translate to the
same amino acid (or to a stop), that translation is done; otherwise,
the codon is translated as X (even if one or more compatible codons
are stops). For example, in the standard code, UAR translates to *
(stop), GGN translates to G (glycine), NNN translates to X, and UGR
translates to X (it could be either a UGA stop or a UGG Trp).

.pp
Degenerate initiation codons are handled essentially the same. If all
codons consistent with the degenerate codon are legal initiators, then
the codon is allowed to initiate a new ORF. Stop codons are never
a legal initiator (not only with 
.I -m 
or
.I -M
but also with the default of allowing any amino acid to initiate),
so degenerate codons consistent with a stop cannot be initiators.
For example, NNN cannot initiate an ORF, nor can UGR -- even
though they translate to X. This means that we don't translate
long stretches of N's as long ORFs of X's, which is probably a
feature, given the prevalence of artificial runs of N's in genome 
sequence assemblies.

.pp
Degenerate DNA codons are not translated to degenerate amino acids
other than X, even when that is possible. For example, SAR and MUH
are decoded as X, not Z (Q|E) and J (I|L). The extra complexity
needed for a degenerate to degenerate translation doesn't seem worthwhile.


.SH OPTIONS

.TP
.B -h
Print brief help. Includes version number and summary of all options. 
Also includes a list of the available
NCBI transl_tables and their numerical codes, for the
.I -c 
option.

.TP
.BI -c " <id>"
Choose alternative genetic code 
.I <id>
where 
.I <id>
is the numerical code of one of the NCBI transl_tables.

.TP
.BI -l " <n>"
Set the minimum reported ORF length to 
.I <n>
aa.

.TP
.B -m
Require ORFs to start with an initiator codon AUG (Met).

.TP
.B -M
Require ORFs to start with an initiator codon, as specified by the
allowed initiator codons in the NCBI transl_table. In the default
Standard code, AUG, CUG, and UUG are allowed as initiators. An 
initiation codon is always translated as Met, even if it does not
normally encode Met as an elongator.

.TP
.B -W
Use a memory-efficient windowed sequence reader.
The default is to read entire DNA sequences into memory, which
may become memory limited for some very large eukaryotic chromosomes.
The windowed reader cannot 
reverse complement a nonrewindable input stream, so 
either
.I <seqfile>
must be a file,
or you must use
.I --watson
to limit translation to the top strand.


.TP
.BI --informat " <s>"
Assert that the input 
.I <seqfile>
is in 
.I <s> 
format. Valid formats include 
FASTA, GenBank, EMBL, Uniprot, DDBJ, Stockholm, Clustal, and Phylip.
.I <s> 
is case-insensitive but must be a complete format name; for example,
.BI --informat " fasta"
or
.BI --informat " embl"
work.

.TP
.B --watson
Only translate the top strand.

.TP
.B --crick
Only translate the bottom strand.


.SH AUTHOR

Easel and its documentation are @EASEL_COPYRIGHT@.
@EASEL_LICENSE@.
See COPYING in the source code distribution for more details.
The Easel home page is: @EASEL_URL@







miniapps/esl-translate.c
/* Translate DNA sequence into six frames, into individual ORFs.
 * 
 */
#include "esl_config.h"

#include "easel.h"
#include "esl_alphabet.h"
#include "esl_gencode.h"
#include "esl_getopts.h"


/*****************************************************************
 * 1. Main loop for reading complete sequences with ReadSeq()
 *****************************************************************/

static int
do_by_sequences(ESL_GENCODE *gcode, ESL_GENCODE_WORKSTATE *wrk, ESL_SQFILE *sqfp)
{
  ESL_SQ *sq = esl_sq_CreateDigital(gcode->nt_abc);
  int     status;

  while (( status = esl_sqio_Read(sqfp, sq )) == eslOK)
    {
      if (sq->n < 3) continue;

      if (wrk->do_watson) {
        esl_gencode_ProcessStart(gcode, wrk, sq);
        esl_gencode_ProcessPiece(gcode, wrk, sq);
        esl_gencode_ProcessEnd(wrk, sq);
      }

      if (wrk->do_crick) {
        esl_sq_ReverseComplement(sq);
        esl_gencode_ProcessStart(gcode, wrk, sq);
        esl_gencode_ProcessPiece(gcode, wrk, sq);
        esl_gencode_ProcessEnd(wrk, sq);
      }

      esl_sq_Reuse(sq);
    }
  if      (status == eslEFORMAT) esl_fatal("Parse failed (sequence file %s)\n%s\n",
					   sqfp->filename, sqfp->get_error(sqfp));     
  else if (status != eslEOF)     esl_fatal("Unexpected error %d reading sequence file %s",
					   status, sqfp->filename);
  
  esl_sq_Destroy(sq);
  return eslOK;
}


static int 
do_by_windows(ESL_GENCODE *gcode, ESL_GENCODE_WORKSTATE *wrk, ESL_SQFILE *sqfp)
{
  ESL_SQ *sq = esl_sq_CreateDigital(gcode->nt_abc);
  int     windowsize  = 4092;                // can be any value, but a multiple of 3 makes most sense. windowsize can be +/-; + means reading top strand; - means bottom strand.
  int     contextsize = 2;                   // contextsize (adjacent window overlap) must be 2, or translation won't work properly.
  int     wstatus;

  ESL_DASSERT1(( windowsize  % 3 == 0 ));  

  while (( wstatus = esl_sqio_ReadWindow(sqfp, contextsize, windowsize, sq)) != eslEOF)
    {
      if (wstatus == eslEOD)
	{
	  if ( (windowsize > 0 && wrk->do_watson) || (windowsize < 0 && wrk->do_crick))
	    esl_gencode_ProcessEnd(wrk, sq);

	  if (windowsize > 0 && ! wrk->do_crick) { esl_sq_Reuse(sq); continue; } // Don't switch to revcomp if we don't need do. Allows -W --watson to work on nonrewindable streams
	  if (windowsize < 0) esl_sq_Reuse(sq);             // Do not Reuse the sq on the switch from watson to crick; ReadWindow needs sq->L
	  windowsize = -windowsize;                         // switch to other strand.
	  continue;
	}
      else if (wstatus == eslEFORMAT) esl_fatal("Parsing failed in sequence file %s:\n%s",          sqfp->filename, esl_sqfile_GetErrorBuf(sqfp));
      else if (wstatus == eslEINVAL)  esl_fatal("Invalid residue(s) found in sequence file %s\n%s", sqfp->filename, esl_sqfile_GetErrorBuf(sqfp));
      else if (wstatus != eslOK)      esl_fatal("Unexpected error %d reading sequence file %s", wstatus, sqfp->filename);

      /* If we're the first window in this input DNA sequence 
       * (or the first window in its revcomp), then initialize.
       * sq->C is the actual context overlap; 0=1st window; 2 (i.e. C)= subsequent.
       */
      if (sq->C == 0) 
	{
	  if (sq->n < 3) continue; // DNA sequence too short; skip it, don't even bother to revcomp, go to next sequence.
	  if ( (windowsize > 0 && wrk->do_watson) || (windowsize < 0 && wrk->do_crick))
	    esl_gencode_ProcessStart(gcode, wrk, sq);
	}

      if ( (windowsize > 0 && wrk->do_watson) || (windowsize < 0 && wrk->do_crick))      
	esl_gencode_ProcessPiece(gcode, wrk, sq);
    }
  esl_sq_Destroy(sq);
  return eslOK;
}


/*****************************************************************
 * x. main() for the esl-translate program
 *****************************************************************/

static ESL_OPTIONS options[] = {
  /* name           type        default  env  range toggles reqs incomp  help                                          docgroup*/
  { "-h",         eslARG_NONE,    FALSE, NULL, NULL, NULL,  NULL, NULL,  "show brief help on version and usage",          0 },
  { "-c",         eslARG_INT,       "1", NULL, NULL, NULL,  NULL, NULL,  "use alt genetic code of NCBI transl table <n>", 0 },
  { "-l",         eslARG_INT,      "20", NULL, NULL, NULL,  NULL, NULL,  "minimum ORF length",                            0 },
  { "-m",         eslARG_NONE,    FALSE, NULL, NULL, NULL,  NULL, "-M",  "ORFs must initiate with AUG only",              0 },
  { "-M",         eslARG_NONE,    FALSE, NULL, NULL, NULL,  NULL, "-m",  "ORFs must start with allowed initiation codon", 0 },
  { "-W",         eslARG_NONE,    FALSE, NULL, NULL, NULL,  NULL, NULL,  "use windowed, memory-efficient seq reading",    0 },
  { "--informat", eslARG_STRING,  FALSE, NULL, NULL, NULL,  NULL, NULL,  "specify that input file is in format <s>",      0 },
  { "--watson",   eslARG_NONE,    FALSE, NULL, NULL, NULL,  NULL, NULL,  "only translate top strand",                     0 },
  { "--crick",    eslARG_NONE,    FALSE, NULL, NULL, NULL,  NULL, NULL,  "only translate bottom strand",                  0 },
  {  0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
};
static char usage[]  = "[-options] <seqfile>";
static char banner[] = "six-frame translation of nucleic acid seq to ORFs";


int
main(int argc, char **argv)
{
  ESL_GETOPTS   *go          = NULL;
  ESL_ALPHABET  *nt_abc      = esl_alphabet_Create(eslDNA);
  ESL_ALPHABET  *aa_abc      = esl_alphabet_Create(eslAMINO);
  ESL_GENCODE   *gcode       = NULL;
  ESL_GENCODE_WORKSTATE *wrk    = NULL;
  char          *seqfile     = NULL;
  int            informat    = eslSQFILE_UNKNOWN;
  ESL_SQFILE    *sqfp        = NULL;
  int            status;
  
  /***************************************************************** 
   * command line parsing
   *****************************************************************/

  go = esl_getopts_Create(options);
  if (esl_opt_ProcessCmdline(go, argc, argv) != eslOK ||
      esl_opt_VerifyConfig(go)               != eslOK)
    {
      printf("Failed to parse command line: %s\n", go->errbuf);
      esl_usage(stdout, argv[0], usage);
      printf("\nTo see more help on available options, do %s -h\n\n", argv[0]);
      exit(1);
    }
  
  if (esl_opt_GetBoolean(go, "-h") )
    {
      esl_banner(stdout, argv[0], banner);
      esl_usage (stdout, argv[0], usage);
      puts("\n where options are:");
      esl_opt_DisplayHelp(stdout, go, /*docgroup=*/0, /*indent=*/2, /*textwidth=*/80);

      puts("\nAvailable NCBI genetic code tables (for -c <id>):");
      esl_gencode_DumpAltCodeTable(stdout);

      exit(0);
    }

  if (esl_opt_ArgNumber(go) != 1) 
    {
      printf("Incorrect number of command line arguments.\n");
      esl_usage(stdout, argv[0], usage);
      printf("\nTo see more help on available options, do %s -h\n\n", argv[0]);
      exit(1);
    }
 
  seqfile = esl_opt_GetArg(go, 1);

  if (esl_opt_IsOn(go, "--informat") &&
      (informat = esl_sqio_EncodeFormat(esl_opt_GetString(go, "--informat"))) == eslMSAFILE_UNKNOWN)
    esl_fatal("%s is not a valid input sequence file format for --informat", esl_opt_GetString(go, "--informat")); 

  status = esl_sqfile_OpenDigital(nt_abc, seqfile, informat, /*env=*/NULL, &sqfp);
  if      (status == eslENOTFOUND) esl_fatal("Failed to find (or open) sequence file %s", seqfile);
  else if (status == eslEFORMAT)   esl_fatal("Failed to recognize format of sequence file %s", seqfile);
  else if (status != eslOK)        esl_fatal("Failure in opening sequence file %s; code %d", seqfile, status);

  /* A limitation. The esl_sqio_ReadWindow() interface needs to use SSI positioning
   * to read reverse complement, and that doesn't work on nonrewindable streams.
   */
  if ( esl_opt_GetBoolean(go, "-W") && ! esl_sqfile_IsRewindable(sqfp) && ! esl_opt_GetBoolean(go, "--watson"))
    esl_fatal("esl-translate can't read reverse complement from a nonrewindable stream (stdin pipe, .gz file, etc).");

  /* Set up the genetic code. Default = NCBI 1, the standard code; allow ORFs to start at any aa
   */
  gcode = esl_gencode_Create(nt_abc, aa_abc);
  esl_gencode_Set(gcode, esl_opt_GetInteger(go, "-c"));  // default = 1, the standard genetic code

  if      (esl_opt_GetBoolean(go, "-m"))   esl_gencode_SetInitiatorOnlyAUG(gcode);
  else if (! esl_opt_GetBoolean(go, "-M")) esl_gencode_SetInitiatorAny(gcode);      // note this is the default, if neither -m or -M are set


  /* Set up the workstate structure, which contains both stateful 
   * info about our position in <sqfp> and the DNA <sq>, as well as
   * one-time config info from options
   */
  wrk = esl_gencode_WorkstateCreate(go, gcode);


  /* The two styles of main processing loop:
   */
  if (esl_opt_GetBoolean(go, "-W"))  do_by_windows(gcode, wrk, sqfp);
  else                               do_by_sequences(gcode, wrk, sqfp);


  esl_gencode_WorkstateDestroy(wrk);
  esl_sqfile_Close(sqfp);
  esl_gencode_Destroy(gcode);
  esl_alphabet_Destroy(aa_abc);
  esl_alphabet_Destroy(nt_abc);
  esl_getopts_Destroy(go);
  return 0;
}




